test/one/profiler/test/SourceLocator.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package one.profiler.test;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+
+public class SourceLocator {
+    public static String tryGetFrom(Throwable e, int ignoreFrames) {
+        StackTraceElement[] stackTrace = e.getStackTrace();
+
+        if (stackTrace.length > ignoreFrames) {
+            StackTraceElement element = stackTrace[ignoreFrames];
+            String className = element.getClassName();
+            String filePath = getFilePath(className);
+
+            return getSourceCodeAt(filePath, element.getLineNumber());
+        }
+        return "No stack trace available";
+    }
+
+    private static String getFilePath(String className) {
+        int dollar = className.lastIndexOf('$');
+        if (dollar >= 0) {
+            className = className.substring(0, dollar);
+        }
+        return "test/" + className.replace('.', '/') + ".java";
+    }
+
+    private static String getSourceCodeAt(String filePath, int lineNumber) {
+        String result = "\t👉 " + filePath + ":" + lineNumber;
+
+        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
+            String line;
+            int currentLine = 1;
+
+            while ((line = reader.readLine()) != null) {
+                if (currentLine == lineNumber) {
+                    return result + "\n\t👉 " + line.trim();
```

Comment:
Can you construct the entire string here? It will be easier to read comparing to when the first part of the message is somewhere else.
Also, let's avoid non-ASCII character. It does not make a big difference here, but who knows where it might go off.

Reply:
I think it is a good idea to keep concerns separate, getting the source code line and building the assertion message are different responsibilities. Renamed methods and classes and further simplified `Assert.java`